### PR TITLE
tetragon: Detect and use disassociate_ctty as exit probe if needed

### PR DESCRIPTION
**Pull request by olsajiri**
Adding support to use disassociate_ctty exit probe in case
acct_process is missing in kernel.

### CODE REVIEW

File: pkg/grpc/tracing/tracing.go
Comment (tixxdz) [inline pkg/grpc/tracing/tracing.go:281]: I think we are using this in other versions... maybe worth a backport? @mtardy @kkourt

File: bpf/process/bpf_exit.c
Comment (tixxdz) [inline bpf/process/bpf_exit.c:46]: While you are it Jiri would it be nice if you can rename it to tg_kp_exit_acct_process and the other one tg_kp_exit_disassociate_ctty ?  I will add later some detect kprobes load / disable policies so just in time for those ;-)  (I can do it later too...)

**Review by tixxdz — APPROVED** [review]
LGTM , thank you Jiri

File: pkg/grpc/tracing/tracing.go
Comment (kkourt) [reply pkg/grpc/tracing/tracing.go:281]: I don't think is worth backporting, but happy to defer to @mtardy

File: pkg/testutils/sensors/load.go
Comment (kkourt) [inline pkg/testutils/sensors/load.go:19]: > At the moment regular expression is implemented via strings.Contains
> call, because it's enough for now.

I think calling it regex is confusing. 

How about have an enum?

```go
type ProgMatch = int
const (
      ProgMatchFull ProgMatch = iota // ==
      ProgMatchPartial // strings.Contains()
)
```

we can add a regex later if needed.

File: pkg/sensors/base/base.go
Comment (kkourt) [inline pkg/sensors/base/base.go:86]: Is there a need to check `has_acct_process` and `has_disassociate_ctty` again?

How about:
```
} else {
   failed = false
}
```

File: pkg/sensors/base/base.go
Comment (kkourt) [inline pkg/sensors/base/base.go:89]: Is there a way to recover from this? If not, I think we should just add a `.Fatal()` here so we die early.

**Review by kkourt — CHANGES_REQUESTED** [review]
Thanks!

I have some comments, but LGTM!

A followup might be a way to test both approaches, since our testing now will only test the `acct_process`. This seems like a `good-first-issue`.

File: bpf/process/bpf_exit.c
Comment (olsajiri) [reply bpf/process/bpf_exit.c:46]: hum, so we want all programs with tg_kp_ prefixes?

File: pkg/testutils/sensors/load.go
Comment (olsajiri) [reply pkg/testutils/sensors/load.go:19]: yep, sounds good, thanks

File: pkg/sensors/base/base.go
Comment (olsajiri) [reply pkg/sensors/base/base.go:86]: yes, I'll change that

File: pkg/sensors/base/base.go
Comment (olsajiri) [reply pkg/sensors/base/base.go:89]: we would start without hooked exit probe.. so it's probably better to bail out then do some strange things

File: bpf/process/bpf_exit.c
Comment (kkourt) [reply bpf/process/bpf_exit.c:46]: I thin `tg_` should suffice. The general idea is to have a (reasonable) way to have a common prefix for tetragon objects so that we can identify them in a system (e.g., when doing `bpftool prog list`).

**Review by kkourt — APPROVED** [review]
Thanks!

**Comment by kkourt** [comment]
The failure is a strange one, but we can ignore it.

image

All tests are successful, but we fail to upload the artifact file (which is empty, because all tests were successful).

**Comment by tixxdz** [comment]
> The failure is a strange one, but we can ignore it.
> 
> image
> 
> All tests are successful, but we fail to upload the artifact file (which is empty, because all tests were successful). @olsajiri feel free to merge.

issue is tracked here https://github.com/cilium/tetragon/issues/1895  cc @mtardy knows about it

File: bpf/process/bpf_exit.c
Comment (tixxdz) [reply bpf/process/bpf_exit.c:46]: So yeh tg_ we do want this, for the extra kp_ up to you no strong opinion, we can go without it.
I'm working on tracing kprobes arm disarm code so being able to identify with tg_ prefix is a quicky way to say if tetragon probes are being disarmed

File: pkg/grpc/tracing/tracing.go
Comment (mtardy) [reply pkg/grpc/tracing/tracing.go:281]: just saw that! indeed, out of curiosity, how you detected that?

**Pull request merged**
